torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,485 @@
+            self._param_rank.clear()
+
+            param_groups = self.partition_parameters()[self.rank]
+            if len(param_groups) == len(self.optim.param_groups) + 1:
+                self.optim.add_param_group(param_groups[-1])
+
+            # Update the bucketing strategy accordingly
+            self._setup_bucket_strategy()
+
+    def consolidate_state_dict(self, recipient_rank: int = 0) -> None:
+        """Update the consolidated state_dict list, one per rank.
+
+        .. warning: This needs to be called on all replicas"""
+
+        # Sync lr and other attributes in case its been updated
+        self._sync_param_groups()
+
+        local_cpu_state = _recursive_copy_to_device(
+            self.local_state_dict(), non_blocking=True, device=torch.device("cpu")
+        )
+
+        # Pull the sharded state from all the other replicas
+        # Store all the states in order, rank by rank
+        if self.rank == recipient_rank:
+            logging.debug("Pulling the sharded optimizer state from all replicas")
+            self._all_states = []
+            for rank in range(self.world_size):
+                if rank == self.rank:
+                    self._all_states.append(local_cpu_state)
+
+                    # Sync with other replicas, almost noop
+                    dist.broadcast_object_list([0], src=self.global_rank, group=self.group)
+                else:
+                    # Fetch the optim state from the other replicas
+                    global_rank = _get_global_rank(self.group, rank)
+                    replica_state = [0]
+                    dist.broadcast_object_list(replica_state, src=global_rank, group=self.group)
+
+                    self._all_states.append(
+                        _recursive_copy_to_device(replica_state[0], non_blocking=True, device=torch.device("cpu"))
+                    )
+        else:
+            # Acknowledge broadcasts, and send this rank's shard when needed
+            # Default to CPU space to gain some memory headroom
+            for rank in range(self.world_size):
+                if rank == self.rank:
+                    # Send the state to the reference replica
+                    dist.broadcast_object_list([local_cpu_state], src=self.global_rank, group=self.group)
+                else:
+                    global_rank = _get_global_rank(self.group, rank)
+
+                    # Discard this tensor/rank, broadcast necessary for syncing and because NCCL does not support gather
+                    dist.broadcast_object_list([0], src=global_rank, group=self.group)
+
+    def partition_parameters(self) -> List[List[Dict]]:
+        """Partitions parameters across distributed data parallel ranks.
+
+        Returns: a list of ``param_groups`` (which is a list of dict) where each
+        element of the list contains the param_groups for a rank. Element 0
+        corresponds to rank 0, etc. We need all the ranks for the broadcast
+        inside ``step()``.
```

Comment:
do we need indent to make the docs render correctly? Feel free to ignore, I can fix doc problem after this PR lands.

Reply:
it works for fairscale, but easier to read with the indent, doing that.